src/persist-client/src/cli/admin.rs

```diff
@@ -273,3 +302,152 @@ pub async fn force_compaction(
         return Ok(());
     }
 }
+
+/// Wrap a lower-level service (Blob or Consensus) to make it read only.
+/// This is probably elaborate enough to work in general -- folks may expect to read
+/// their own writes, among other things -- but it should handle the case of GC, where
+/// all reads finish before the writes begin.
+#[derive(Debug)]
+struct ReadOnly<T>(T);
+
+#[async_trait]
+impl Blob for ReadOnly<Arc<dyn Blob + Sync + Send>> {
```

Comment:
wow, I love this. I honestly wouldn't mind going further and making it an input to `PersistConfig` so we can return a `ReadOnlyBlob` and `ReadOnlyConsensus` from `(Blob|Consensus)Config::open` directly. Then we can easily apply this to all admin/inspect commands. This is a really nice addition to protect ourselves from accidentally corrupting prod state

Reply:
(this doesn't need to be done in this PR, but overall huge thumbs up to read-only blob/consensus!)

Reply:
I'm also hugely in favor of this, tho my personal instinct is to a) leave it as a cli-only thing for now (because of the below) and b) (maybe) panic instead of no-op for the calls that mutate. (b) would mean an invocation of `persistcli admin gc` without `--commit` would panic, but maybe that's better than the confusion that could happen as a result of silent no-ops? I guess it could also sniff those panics somehow and turn them into a nice message.

I'd love for the long term answer here to be something like the [Blob vs BlobRead split in old persist](https://github.com/MaterializeInc/materialize/pull/9574), but that would require "pushing" the dyn from Blob+Consensus to something like `dyn Machine<K, V, T, D>`. The latter seems desirable, anyway, as a way to split a bunch of persist-client code out into a crate that's only referenced at the leaves of the code graph

Reply:
I still think the silent no-ops are a pretty big footgun. At the very least, each of the no-op methods should log at warn level. It'd be spammy in the output of the admin tool, but that feels better to me than silent no-ops

Reply:
Cool, I'm fine with warning!

If you're worried it will be too spammy, we could have it warn only the first time a method is invoked, with the comment that any data read after that point should be treated with suspicion.

Reply:
given that this is an admin commandline, I'm not super concerned about the spam (could totally turn out that I'm wrong in practice, but shrug). it also seems nice for the --no-commit version of the tool to print out the entirety of the mutations it would have done with --commit

Reply:
+1 a warn log on every call seems pretty reasonable. I'd rather know too much than too little when trying to understand what the tool is going to do